<a href="https://colab.research.google.com/github/itsZENR/Binance/blob/main/%D0%A0%D0%B0%D0%B1%D0%BE%D1%82%D0%B0_%D1%81_%D0%B4%D0%B0%D1%82%D0%B0%D1%81%D0%B5%D1%82%D0%BE%D0%BC_Binance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#@title Библиотеки
import pandas as pd
import numpy as np
import requests
import csv
import datetime
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from datetime import datetime

In [27]:
BINANCE_CSV_1h_ARBUSDT = "/content/binance_quotes_1h_ARBUSDT.csv"
BINANCE_CSV_1h_BELUSDT = "/content/binance_quotes_1h_BELUSDT.csv"
BINANCE_CSV_1h_BTCUSDT = "/content/binance_quotes_1h_BTCUSDT.csv"
BINANCE_CSV_1h_ETHUSDT = "/content/binance_quotes_1h_ETHUSDT.csv"
BINANCE_CSV_1h_XRPUSDT = "/content/binance_quotes_1h_XRPUSDT.csv"

In [45]:
# Загружаем CSV-файл с данными и преобразуем данные в DataFrame
df1 = pd.read_csv(BINANCE_CSV_1h_ARBUSDT)[['open', 'close']]
df2 = pd.read_csv(BINANCE_CSV_1h_BELUSDT)[['open', 'close']]
df3 = pd.read_csv(BINANCE_CSV_1h_BTCUSDT)[['open', 'close']]
df4 = pd.read_csv(BINANCE_CSV_1h_ETHUSDT)[['open', 'close']]
df5 = pd.read_csv(BINANCE_CSV_1h_XRPUSDT)[['open', 'close']]
df1

open   close
0    0.5000  1.3250
1    1.3251  1.3040
2    1.3040  1.2890
3    1.2890  1.3273
4    1.3274  1.4300
..      ...     ...
542  1.6401  1.6591
543  1.6590  1.6450
544  1.6450  1.6723
545  1.6725  1.6621
546  1.6621  1.6720

[547 rows x 2 columns]

In [49]:
for j in [df1, df2, df3, df4, df5]:
  for i in range(1, 100):
    j[f'SMA{i}'] = j['close'].rolling(window=i).mean()
df1

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


open   close    SMA1     SMA2      SMA3      SMA4     SMA5      SMA6  \
0    0.5000  1.3250  1.3250      NaN       NaN       NaN      NaN       NaN   
1    1.3251  1.3040  1.3040  1.31450       NaN       NaN      NaN       NaN   
2    1.3040  1.2890  1.2890  1.29650  1.306000       NaN      NaN       NaN   
3    1.2890  1.3273  1.3273  1.30815  1.306767  1.311325      NaN       NaN   
4    1.3274  1.4300  1.4300  1.37865  1.348767  1.337575  1.33506       NaN   
..      ...     ...     ...      ...       ...       ...      ...       ...   
542  1.6401  1.6591  1.6591  1.64955  1.645633  1.649300  1.64830  1.647500   
543  1.6590  1.6450  1.6450  1.65205  1.648033  1.645475  1.64844  1.647750   
544  1.6450  1.6723  1.6723  1.65865  1.658800  1.654100  1.65084  1.652417   
545  1.6725  1.6621  1.6621  1.66720  1.659800  1.659625  1.65570  1.652717   
546  1.6621  1.6720  1.6720  1.66705  1.668800  1.662850  1.66210  1.658417   

         SMA7      SMA8  ...     SMA90     SMA91     SMA92     SMA93  \
0         NaN       NaN  ...       NaN       NaN       NaN       NaN   
1         NaN       NaN  ...       NaN       NaN       NaN       NaN   
2         NaN       NaN  ...       NaN       NaN       NaN       NaN   
3         NaN       NaN  ...       NaN       NaN       NaN       NaN   
4         NaN       NaN  ...       NaN       NaN       NaN       NaN   
..        ...       ...  ...       ...       ...       ...       ...   
542  1.654786  1.658725  ...  1.351432  1.350093  1.348861  1.347619   
543  1.647143  1.653563  ...  1.356086  1.354658  1.353299  1.352045   
544  1.651257  1.650287  ...  1.360980  1.359560  1.358111  1.356729   
545  1.653800  1.652613  ...  1.365850  1.364289  1.362849  1.361380   
546  1.655471  1.656075  ...  1.370752  1.369214  1.367634  1.366173   

        SMA94     SMA95     SMA96     SMA97     SMA98     SMA99  
0         NaN       NaN       NaN       NaN       NaN       NaN  
1         NaN       NaN       NaN       NaN       NaN       NaN  
2         NaN       NaN       NaN       NaN       NaN       NaN  
3         NaN       NaN       NaN       NaN       NaN       NaN  
4         NaN       NaN       NaN       NaN       NaN       NaN  
..        ...       ...       ...       ...       ...       ...  
542  1.346427  1.345320  1.344222  1.343031  1.341973  1.340942  
543  1.350783  1.349569  1.348442  1.347323  1.346112  1.345034  
544  1.355452  1.354167  1.352931  1.351780  1.350639  1.349407  
545  1.359978  1.358680  1.357375  1.356119  1.354947  1.353785  
546  1.364684  1.363262  1.361944  1.360619  1.359342  1.358149  

[547 rows x 101 columns]

In [51]:
df1 = df1.iloc[:, 1:]
df2 = df2.iloc[:, 1:]
df3 = df3.iloc[:, 1:]
df4 = df4.iloc[:, 1:]
df5 = df5.iloc[:, 1:]
df1

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


close    SMA1     SMA2      SMA3      SMA4     SMA5      SMA6      SMA7  \
0    1.3250  1.3250      NaN       NaN       NaN      NaN       NaN       NaN   
1    1.3040  1.3040  1.31450       NaN       NaN      NaN       NaN       NaN   
2    1.2890  1.2890  1.29650  1.306000       NaN      NaN       NaN       NaN   
3    1.3273  1.3273  1.30815  1.306767  1.311325      NaN       NaN       NaN   
4    1.4300  1.4300  1.37865  1.348767  1.337575  1.33506       NaN       NaN   
..      ...     ...      ...       ...       ...      ...       ...       ...   
542  1.6591  1.6591  1.64955  1.645633  1.649300  1.64830  1.647500  1.654786   
543  1.6450  1.6450  1.65205  1.648033  1.645475  1.64844  1.647750  1.647143   
544  1.6723  1.6723  1.65865  1.658800  1.654100  1.65084  1.652417  1.651257   
545  1.6621  1.6621  1.66720  1.659800  1.659625  1.65570  1.652717  1.653800   
546  1.6720  1.6720  1.66705  1.668800  1.662850  1.66210  1.658417  1.655471   

         SMA8      SMA9  ...     SMA90     SMA91     SMA92     SMA93  \
0         NaN       NaN  ...       NaN       NaN       NaN       NaN   
1         NaN       NaN  ...       NaN       NaN       NaN       NaN   
2         NaN       NaN  ...       NaN       NaN       NaN       NaN   
3         NaN       NaN  ...       NaN       NaN       NaN       NaN   
4         NaN       NaN  ...       NaN       NaN       NaN       NaN   
..        ...       ...  ...       ...       ...       ...       ...   
542  1.658725  1.661689  ...  1.351432  1.350093  1.348861  1.347619   
543  1.653563  1.657200  ...  1.356086  1.354658  1.353299  1.352045   
544  1.650287  1.655644  ...  1.360980  1.359560  1.358111  1.356729   
545  1.652613  1.651600  ...  1.365850  1.364289  1.362849  1.361380   
546  1.656075  1.654767  ...  1.370752  1.369214  1.367634  1.366173   

        SMA94     SMA95     SMA96     SMA97     SMA98     SMA99  
0         NaN       NaN       NaN       NaN       NaN       NaN  
1         NaN       NaN       NaN       NaN       NaN       NaN  
2         NaN       NaN       NaN       NaN       NaN       NaN  
3         NaN       NaN       NaN       NaN       NaN       NaN  
4         NaN       NaN       NaN       NaN       NaN       NaN  
..        ...       ...       ...       ...       ...       ...  
542  1.346427  1.345320  1.344222  1.343031  1.341973  1.340942  
543  1.350783  1.349569  1.348442  1.347323  1.346112  1.345034  
544  1.355452  1.354167  1.352931  1.351780  1.350639  1.349407  
545  1.359978  1.358680  1.357375  1.356119  1.354947  1.353785  
546  1.364684  1.363262  1.361944  1.360619  1.359342  1.358149  

[547 rows x 100 columns]

In [61]:
def SMA_test(df):
  for j in range(1, 100):
    predict = []
    target = []
    SMA = []
    procent = []
    for i in range(len(df)-1):
      # Если нынешняя цена меньше цены будущей
      if df.iloc[i, 0] < df.iloc[i+1, 0]:
        target.append(1)
      else:
        target.append(-1)
      
      if df.iloc[i, 0] < df.iloc[i, j]:
        predict.append(1)
      else:
        predict.append(-1)

    #  Считает скольок рах мы угадали
    sum = 0
    for i in range(len(target)):
      if target[i] == predict[i]:
        sum+=1
    print('на', df.iloc[:, j].name, 'всего', len(target), 'из них угадано', sum)
    SMA.append(df.iloc[:, j].name)
    print('%', sum/len(target)*100)
    procent.append(sum/len(target)*100)
    print()

In [62]:
for i, j in enumerate([df1, df2, df3, df4, df5]):
  print('номер датасета', i+1)
  SMA_test(j)
  print('-'*100)

номер датасета 1
на SMA1 всего 546 из них угадано 270
% 49.45054945054945

на SMA2 всего 546 из них угадано 284
% 52.01465201465202

на SMA3 всего 546 из них угадано 291
% 53.2967032967033

на SMA4 всего 546 из них угадано 283
% 51.83150183150184

на SMA5 всего 546 из них угадано 287
% 52.56410256410257

на SMA6 всего 546 из них угадано 284
% 52.01465201465202

на SMA7 всего 546 из них угадано 292
% 53.47985347985348

на SMA8 всего 546 из них угадано 293
% 53.663003663003664

на SMA9 всего 546 из них угадано 291
% 53.2967032967033

на SMA10 всего 546 из них угадано 294
% 53.84615384615385

на SMA11 всего 546 из них угадано 294
% 53.84615384615385

на SMA12 всего 546 из них угадано 295
% 54.02930402930403

на SMA13 всего 546 из них угадано 286
% 52.38095238095239

на SMA14 всего 546 из них угадано 291
% 53.2967032967033

на SMA15 всего 546 из них угадано 280
% 51.28205128205128

на SMA16 всего 546 из них угадано 282
% 51.64835164835166

на SMA17 всего 546 из них угадано 284
% 52.0146520

df1
на SMA12 всего 546 из них угадано 295 
% 54.02930402930403

df2
на SMA8 всего 22558 из них угадано 11815
% 52.37609717173508

df3
на SMA4 всего 49411 из них угадано 27144
% 54.935135900912755

df4
на SMA4 всего 49411 из них угадано 26842
% 54.32393596567566

df5
на SMA6 всего 43232 из них угадано 23396
% 54.117320503330866